components/resource_metering/src/reporter/pubsub.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::DataSinkRegHandle;
+use crate::error::Result;
+use crate::metrics::{IGNORED_DATA_COUNTER, REPORT_DATA_COUNTER, REPORT_DURATION_HISTOGRAM};
+use crate::DataSink;
+
+use std::sync::Arc;
+
+use futures::channel::mpsc::{channel, Sender};
+use futures::{SinkExt, StreamExt};
+use grpcio::{RpcContext, ServerStreamingSink, WriteFlags};
+use kvproto::resource_usage_agent::{ResourceMeteringRequest, ResourceUsageRecord};
+use kvproto::resource_usage_agent_grpc::ResourceMeteringPubSub;
+use tikv_util::{info, warn};
+
+/// `PubSubService` implements [ResourceMeteringPubSub].
+///
+/// If a client subscribes to resource metering records, the `PubSubService` is responsible for
+/// registering them to the reporter. Then the reporter sends data to the client periodically.
+///
+/// [ResourceMeteringPubSub]: kvproto::resource_usage_agent_grpc::ResourceMeteringPubSub
+#[derive(Clone)]
+pub struct PubSubService {
+    data_sink_reg_handle: DataSinkRegHandle,
+}
+
+impl PubSubService {
+    pub fn new(data_sink_reg_handle: DataSinkRegHandle) -> Self {
+        Self {
+            data_sink_reg_handle,
+        }
+    }
+}
+
+impl ResourceMeteringPubSub for PubSubService {
+    fn subscribe(
+        &mut self,
+        ctx: RpcContext<'_>,
+        _: ResourceMeteringRequest,
+        mut sink: ServerStreamingSink<ResourceUsageRecord>,
+    ) {
+        info!("accept a new subscriber"; "from" => ?ctx.peer());
+
+        // The `tx` is for the reporter and the `rx` is for the gRPC stream sender.
+        //
+        // The reporter calls `tx.try_send` roughly every minute. If the the gRPC
+        // stream sender does not send data out over the network in time, it discards
+        // the incoming records to prevent memory overflow.
+        let (tx, mut rx) = channel(1);
+
+        let data_sink = DataSinkImpl { tx };
+        let handle = self.data_sink_reg_handle.register(Box::new(data_sink));
+
+        let report_task = async move {
+            let _h = handle;
+
+            loop {
+                let records = rx.next().await;
+                if records.is_none() {
```

Comment:
When will this happen?

Reply:
When datasink is dropped, then tx is dropped, and the channel becomes closed, the routine comes here.